KAFKA-5280: Protect txn metadata map with read-write lock

guozhangwang opened the pull request:
Two major changes plus one minor change:

0. change stateLock to a read-write lock.

1. Put the check of "isCoordinator" and "coordinatorLoading" together with the return of the metadata, under one read lock block, since otherwise we can get incorrect behavior if there is a change in the metadata cache after the check but before the accessing of the metadata.

2. Grab the read lock right before trying to append to local txn log, and until the local append returns; this is to avoid the scenario that the epoch has actually changed when we are appending to local log (e.g. emigration followed by immigration).

3. only watch on txnId instead of txnId and txnPartitionId in the txn marker purgatory, and disable reaper thread, as we can now safely clear all the delayed operations by traversing the marker queues.

guozhangwang committed:
bump up producer epoch

guozhangwang committed:
Jason's comments

guozhangwang committed:
Jun's comments

guozhangwang committed:
bump up limit for class fanout to 40 for Sender class

guozhangwang committed:
change the state lock to read-write lock

guozhangwang committed:
Jun's comments round two

guozhangwang committed:
grab the read lock until append to local is done

guozhangwang committed:
rebased from trunk

guozhangwang committed:
put the validation and return of metadata under the same lock

guozhangwang committed:
rebase from trunk

guozhangwang committed:
fix unit tests

guozhangwang commented:
ping @dguy @hachikuji @apurvam @junrao for reviews.

Two major changes plus one minor change:

0) change `stateLock` to a read-write lock.

1) Put the check of "isCoordinator" and "coordinatorLoading" together with the return of the metadata, under one read lock block, since otherwise we can get incorrect behavior if there is a change in the metadata cache after the check but before the accessing of the metadata.

2) Grab the read lock right before trying to append to local txn log, and until the local append returns; this is to avoid the scenario that the epoch has actually changed when we are appending to local log (e.g. emigration followed by immigration).

3) Minor: enable reaper thread in the txn marker purgatory, since we are watching each delayed operation on two keys: txnId and partitionId, we need the reaper to purge these already completed items even though nothing will be expired.

guozhangwang commented:
One thing I realized while working on this PR is that, in purgatory's `cancelForKey` function, we only remove the elements from the watch-list but does not set its completed boolean; so if it is watched on other keys they will not be purged. I think this is a bug, that we should try-set its `completed` bit upon canceling it from `cancelForKey`?

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -98,18 +100,16 @@ class TransactionCoordinator(brokerId: Int,
       // and return a new producerId from the producerId manager
       val producerId = producerIdManager.generateProducerId()
       responseCallback(InitProducerIdResult(producerId, producerEpoch = 0, Errors.NONE))
-    } else if (!txnManager.isCoordinatorFor(transactionalId)) {
```
Hmm.. why don't we need this anymore?

hachikuji replied:
Never mind, I see we moved the check into the state manager.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -48,8 +48,10 @@ object TransactionCoordinator {
     val producerIdManager = new ProducerIdManager(config.brokerId, zkUtils)
+    // we need to enable reaper thread still even with infinite delays for txn marker purgatory, since we
+    // are still relying on the thread to remove already completed tasks as each task is watched on two lists: txn id and partition id
+    val txnMarkerPurgatory = DelayedOperationPurgatory[DelayedTxnMarker]("txn-marker-purgatory", config.brokerId)
```
With this, should we just remove reaperEnabled in the constructor of DelayedOperationPurgatory since there won't be a use case for disabling the reaper?

guozhangwang replied:
You mean in the `object DelayedOperationPurgatory`'s apply function? Yes we can.

junrao replied:
I meant if we can just get rid of reaperEnabled in class DelayedOperationPurgatory.

guozhangwang replied:
It is still used in some unit tests to not enable the reaper thread.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -91,9 +89,9 @@ class TransactionStateManager(brokerId: Int,
   def transactionsToExpire(): Iterable[TransactionalIdAndProducerIdEpoch] = {
     val now = time.milliseconds()
-    transactionMetadataCache.flatMap { case (_, entry) =>
+    transactionMetadataCache.flatMap { case (txnPartition, entry) =>
```
Don't we need the read lock for this?

guozhangwang replied:
As commented above, this is just to generate a "candidate list" of txns that can be expired, they may not contain all the validate txns nor does every txn included here is expirable. That is why we are not returning the object itself but just the `pid, txnid, epoch` fields and then tries to get the object iteratively again in the `transaction coordinator`.

guozhangwang replied:
EDIT: actually we do need the read lock not for correctness but to avoid ConcurrentModificationException.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -91,9 +89,9 @@ class TransactionStateManager(brokerId: Int,
   def transactionsToExpire(): Iterable[TransactionalIdAndProducerIdEpoch] = {
     val now = time.milliseconds()
-    transactionMetadataCache.flatMap { case (_, entry) =>
+    transactionMetadataCache.flatMap { case (txnPartition, entry) =>
```
Hmm, it may not be safe to iterate a map that could be updated concurrently since the iterator will throw a ConcurrentModificationException. So, it's better to protect this under the read lock.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,32 +111,52 @@ class TransactionStateManager(brokerId: Int,
+   * Get the transaction metadata associated with the given transactional id, or None if not found, or an error if
+   * the coordinator does not own the transaction partition or is still loading it
+   *
+   * Need to grab the state lock before calling this function
```
Is this comment accurate since the method itself already grabs the lock?

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,32 +111,52 @@ class TransactionStateManager(brokerId: Int,
+        }
+      case None =>
+        Left(Errors.NOT_COORDINATOR)
       }
```
nit: this is no longer aligned

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,32 +111,52 @@ class TransactionStateManager(brokerId: Int,
+    // we only need to check loading partition set but not leaving partition set here since there are three possible cases:
+    //    1) it is not in the leaving partitions set, hence safe to return COORDINATOR_LOAD_IN_PROGRESS
+    //    2) it is in the leaving partitions with a smaller epoch than the latest loading epoch, hence safe to return COORDINATOR_LOAD_IN_PROGRESS
+    //    3) it is in the leaving partition with a larger epoch, return true is also OK since the client will then retry
```
What does "return true" mean in this context?

guozhangwang replied:
removed the comment.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,32 +111,52 @@ class TransactionStateManager(brokerId: Int,
+          case null =>
+            Right(None)
+          case txnMetadata =>
+            // we only need to check leaving partition set but not loading partition set here since there are three possible cases:
```
Is this comment still needed? We already checked the loading partition set above and we are holding the lock.

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,32 +111,52 @@ class TransactionStateManager(brokerId: Int,
+   *
+   * Need to grab the state lock before calling this function
    */
   def addTransaction(transactionalId: String, txnMetadata: TransactionMetadata): CoordinatorEpochAndTxnMetadata = {
```
Don't we need the read lock when `transactionMetadataCache.get` below? Also, why is this operation safe without checking the loading/leaving partitions but `getTransactionState` is not?

guozhangwang replied:
We do need a lock.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -484,13 +485,42 @@ class TransactionStateManager(brokerId: Int,
+    // note that we need to grab the read lock and double check the metadata and coordinator epoch again
+    // until we have appended to log to avoid the scenario that an emigration and immigration happens after
+    // the check when handling the request
+    inReadLock(stateLock) {
```
Do we need to hold the read lock while doing the `append` below? Note that `getTransactionState` already acquires it.

guozhangwang replied:
We do. The scenario trying to avoid here is that 1) get the metadata with read lock, 2) this thread gets suspended on the object or the cache map itself gets modified, 3) append to local log which is incorrect.

guozhangwang replied:
> One thing I realized while working on this PR is that, in purgatory's cancelForKey function, we only remove the elements from the watch-list but does not set its completed boolean; so if it is watched on other keys they will not be purged. I think this is a bug, that we should try-set its completed bit upon canceling it from cancelForKey?

What do you think @junrao @hachikuji ?

junrao replied:
Another way to do this is to have DelayedTxnMarker.tryComplete() check the expected coordinator epoch and transaction state and error out if they don't match. This way, we just need to replace cancelForKey() calls with checkAndComplete() and get rid of cancelForKey().

guozhangwang replied:
The problem with that is because we have already check that in `handler`, and re-checking it again in the delayed operation sounds like an overkill to me.

I instead made the changes to only watch on transactional id, not txn partition id, as in the makers queue we already maintain the markers per txn partition so that we can call `cancel` on the transactional ids of that txn partition anyways. WDYT?

hachikuji reviewed (commented):
A few comments in TransactionStateManager.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,32 +111,52 @@ class TransactionStateManager(brokerId: Int,
+          case txnMetadata =>
+            // we only need to check leaving partition set but not loading partition set here since there are three possible cases:
+            //    1) it is not in the loading partitions set, hence safe to return NONE
+            //    2) it is in the loading partitions with a smaller epoch, hence safe to return NONE
```
Hmm, if it is loading, it seems that in the check in line 129, we return COORDINATOR_LOAD_IN_PROGRESS instead of NONE? Since we are checking leavingPartitions anyway, perhaps it's better to just check it after line 130 after we check loadingPartitions? Then, the comments can probably be combined and simplified.

guozhangwang committed:
rebase from trunk

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,32 +111,52 @@ class TransactionStateManager(brokerId: Int,
    * Add a new transaction metadata, or retrieve the metadata if it already exists with the associated transactional id
    * along with the current coordinator epoch for that belonging transaction topic partition
+   *
+   * Need to grab the state lock before calling this function
```
Hmm, ideally, it seems that addTransaction() should also hold onto the read lock when accessing transactionMetadataCache to prevent concurrent coordinator change. One way to do that is to change getTransactionState to sth like getMaybCreateTxnMetadata(), which returns a TxnMetadata if it exists, but can optionally create a new TxnMetadata if it's missing and creation is enabled. Only initPid will enable the creation.

junrao reviewed (commented):
@guozhangwang : Thanks for the patch. A few comments below.

guozhangwang commented:
@hachikuji @junrao I addressed your comments with a single function, one small drawback is that if there is already an existing txn metadata upon `initPid` it will waste a pid. But in practice producer fencing should be rare so hopefully it does not have huge impact.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -484,13 +485,42 @@ class TransactionStateManager(brokerId: Int,
+              // do not need to check the metadata object itself since no current thread should be able to modify it
+              // under the same coordinator epoch, so directly append to txn log now
+
+              replicaManager.appendRecords(
```
We actually have to be a bit careful not to introduce similar deadlocks that we had in GroupManager before. It seems a deadlock could happen in the following 2 code paths.

1. Here, we hold a lock on txnMetadata when calling replicaManager.appendRecords(), which will eventually call ProducerPurgatory.tryCompleteElseWatch() which holds a lock on the delayed produce when calling safeTryComplete(). 

2. A fetch follower request will call DelayedOperationPurgatory.checkAndComplete(), which holds a lock on the delayed produce and calls safeTryComplete(), which then holds a lock on txnMetadata.

If the caller of 1 and 2 each first holds the first lock and then proceeds, a deadlock will happen.

guozhangwang replied:
Hmm, it's a bit tricky. We need to hold the lock until local append returns, but not when calling `delayedProducePurgatory.tryCompleteElseWatch`.

BTW I have been thinking about this race condition a bit more, and now I feel it is actually okay to not lock the memory cache while appending to local log, and here is my reasoning: When there is an emigration / immigration happening after the txn metadata object is checked but before the local log is appended, then the following case can happen:

1. `handleFollower` is received and handled, in which the local log append will be truncated; the delayed produce will eventually be timed out and an error returned to the user.
2. `handleFollower` is not received, and `handleLeader` with new epoch is received and handled; the append log will be appended and epoch check will fail in the callback, and a `NOT_COORDINATOR` will be sent back to client and it will retry; we will eventually get an idempotent append in the txn log which is okay for loading.
3. none of `handleFollower` or `handleLeader` is received, and the append is successful, it is locally the same as the event happens after the append to the txn log.

WDTY?

junrao replied:
I thought we need to hold onto the txnMetadata lock since a transaction could potentially be expired in parallel?

Jason had the idea that transaction coordinator could potentially pass in a customized DelayedProduce that implements a customized safeComplete() that doesn't sync on the delayedProduce, but on txnMetadata, to avoid the potential deadlock.

guozhangwang replied:
Sorry my previous comment is a bit confusing, it actually contains two parts:

1. we need to hold on the `txnMetadata` object lock, which is a bit tricky to implement; and what @hachikuji suggested sounds promising to me.

2. we do NOT need to hold on the read-write lock on the cache map; this is what I meant in the BTW paragraph.

junrao commented:
Also, TransactionMarkerChannelManager.txnMarkerPurgatory is only used in TransactionMarkerChannelManager, but is passed to TransactionCoordinator for shutting down. Could we just shut down txnMarkerPurgatory in TransactionMarkerChannelManager and avoid passing txnMarkerPurgatory to TransactionCoordinator?

guozhangwang committed:
fix unit tests

guozhangwang commented:
> Also, TransactionMarkerChannelManager.txnMarkerPurgatory is only used in

ack.

guozhangwang committed:
remove cache lock during appending

guozhangwang committed:
address Jun's comment round three

hachikuji commented on core/src/main/scala/kafka/server/DelayedProduce.scala:
```diff
@@ -124,6 +124,19 @@ class DelayedProduce(delayMs: Long,
   }
 }
 
+class SafeDelayedProduce(delayMs: Long,
```
Actually I wonder if we need an extension. Perhaps the original `DelayedProduce` could take an optional lock in the constructor which defaults to `this` or just `new Object`?

guozhangwang replied:
I do not have a strong preference in either case, since we need to modify `replica manager` to pass an additional parameter anyways. @hachikuji @junrao Let me know if you have a strong opinion in this (fewer virtual calls?) and I can change.

BTW any other comments besides this one? Just want to cumulate more comments before I rebase again..

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -48,11 +48,13 @@ object TransactionCoordinator {
 
     val producerIdManager = new ProducerIdManager(config.brokerId, zkUtils)
+    // we need to enable reaper thread still even with infinite delays for txn marker purgatory, since we
+    // are still relying on the thread to remove already completed tasks as each task is watched on two lists: txn id and partition id
```
The comment seems inaccurate since we only watch on transactional id now? Also, would it be better to construct txnMarkerPurgatory inside TransactionMarkerChannelManager?

guozhangwang replied:
Updated the comments.

The reason we did not move it inside `TransactionMarkerChannelManager` is for convenient unit tests; otherwise we need to expose the purgatory in order to check its status.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -348,10 +334,15 @@ class TransactionCoordinator(brokerId: Int,
-              val preSendResult: Either[Errors, (TransactionMetadata, TransactionMetadataTransition)] = txnManager.getTransactionState(transactionalId) match {
-                case Some(epochAndMetadata) =>
-                  if (epochAndMetadata.coordinatorEpoch == coordinatorEpoch) {
+              // we do NOT need to grab the read lock until we have completed putting into the sender queue,
```
The comment is not super clear to me. The read lock is only in TransactionStateManager. So it's a bit weird to reference it here.

guozhangwang replied:
Okay, I will remove them.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -70,10 +70,8 @@ class TransactionStateManager(brokerId: Int,
-  // by the read lock
-  /** lock protecting access to loading and owned partition sets */
-  private val stateLock = new ReentrantLock()
+  /** lock protecting access to the metadata cache, including loading and leaving partition sets */
```
To be clear, would it be better to say "transactional metadata cache"?

guozhangwang replied:
ack

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -87,13 +85,16 @@ class TransactionStateManager(brokerId: Int,
   private val transactionTopicPartitionCount = getTransactionTopicPartitionCount
 
-  // this is best-effort expiration and hence not grabing the lock on metadata upon checking its state
+  // this is best-effort expiration and hence not grabbing the lock on metadata upon checking its state
```
This comment is no longer accurate since we are holding on to the read lock now?

guozhangwang replied:
The original comment is for "the metadata object lock" actually, I will make it more clear.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,47 +115,44 @@ class TransactionStateManager(brokerId: Int,
+                Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata)))
+              } else {
+                Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata)))
+              }
```
It seems that we can always return Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata)))?

guozhangwang replied:
Ack.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,47 +115,44 @@ class TransactionStateManager(brokerId: Int,
    */
-  def getTransactionState(transactionalId: String): Option[CoordinatorEpochAndTxnMetadata] = {
+  def getAndMaybeAddTransactionState(transactionalId: String,
+                                     txnMetadata: TransactionMetadata = null): Either[Errors, Option[CoordinatorEpochAndTxnMetadata]]
```
txnMetadata => newlyCreatedtxnMetadata ? Perhaps also add a comment to explain how it's used. Finally, would it be better to make it an Option?

guozhangwang replied:
Ack.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,47 +115,44 @@ class TransactionStateManager(brokerId: Int,
-  def addTransaction(transactionalId: String, txnMetadata: TransactionMetadata): CoordinatorEpochAndTxnMetadata = {
-    val partitionId = partitionFor(transactionalId)
+    if (leavingPartitions.exists(_.txnPartitionId == partitionId))
+      Right(None)
```
It seems that we should return Left(Errors.NOT_COORDINATOR) in this case?

guozhangwang replied:
Ahh right.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -113,47 +115,44 @@ class TransactionStateManager(brokerId: Int,
+                Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata)))
+              }
+            } else {
+              Right(None)
```
It seems that we should return Left(INVALID_TXN_STATE)? If so, it seems that this method will never return Right(None) and the caller should be adjusted.

guozhangwang replied:
I don't think so: `None` is used to indicate there is no txn metadata and the caller is not trying to create a new one, different callers will handle `None` differently, e.g. return `NOT_COORDINATOR`, return `INVALID_PID`, throw an exception.

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -484,13 +466,38 @@ class TransactionStateManager(brokerId: Int,
+            // the coordinator epoch has changed, reply to client immediately with with NOT_COORDINATOR
+            responseCallback(Errors.NOT_COORDINATOR)
+          } else {
+            // do not need to check the metadata object itself since no current thread should be able to modify it
```
current => concurrent? Also, is it really safe not to hold a read lock of stateLock during log append? Immediately after the check in line 481, an emmigration followed by an immigration could have completed. Since appendRecords() is only synchronized on a txnMetadata which is now obsolete, entries with a high coordinator epoch could have been appended to the log. So, appending entries for an old coordinator epoch would make the log in a bad state.

guozhangwang replied:
About the locking, here is my reasoning: https://github.com/apache/kafka/pull/3082#discussion_r117591792

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -336,7 +336,8 @@ class ReplicaManager(val config: KafkaConfig,
                     entriesPerPartition: Map[TopicPartition, MemoryRecords],
-                    responseCallback: Map[TopicPartition, PartitionResponse] => Unit) {
+                    responseCallback: Map[TopicPartition, PartitionResponse] => Unit,
+                    delayedProduceSyncObject: Object = null) {
```
Perhaps add a comment describing what delayedProduceSyncObject is for? Perhaps it's better to make delayedProduceSyncObject an Option?

hachikuji replied:
Maybe a slightly nicer way to do this is to pass a `DelayedProduce` builder. For example:

`delayedProduceBuilder: (timeout, produceMetadata) -> DelayedProduce`

Then we can remove the `responseCallback` argument since this can be added to the DelayedProduce in the builder.

junrao commented on core/src/main/scala/kafka/server/DelayedOperation.scala:
```diff
@@ -253,6 +252,9 @@ final class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: Stri
 
+  /**
+    * Cancel watching on any delayed operations for the given key. Note the operation will not be completed
+    */
```
Hmm, it seems that there is still a problem with an infinite timeout in DelayedTxnMarker. The issue is that the operation will be added to DelayedOperationPurgatory.timeoutTimer, but will never be removed, which leads to memory leak.

guozhangwang replied:
It will call `TimeTask.cancel()` which will remove itself from the timer.

junrao commented on core/src/main/scala/kafka/server/DelayedProduce.scala:
```diff
@@ -124,6 +124,19 @@ class DelayedProduce(delayMs: Long,
+                         produceMetadata: ProduceMetadata,
+                         replicaManager: ReplicaManager,
+                         responseCallback: Map[TopicPartition, PartitionResponse] => Unit)
+  extends DelayedProduce(delayMs, produceMetadata, replicaManager, responseCallback) {
```
Could we add a comment on when SafeDelayedProduce is needed (e.g. avoid deadlock)?

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala:
```diff
@@ -142,11 +141,13 @@ class TransactionMarkerChannelManager(config: KafkaConfig,
 
   def shutdown(): Unit = {
+    // wake up the thread in case it is blocked inside poll
+    networkClient.wakeup()
```
We don't want txnMarkerSendThread to process any new requests after networkClient is waking up. So, a better shutdown sequence is probably:

txnMarkerSendThread.initiateShutdown()
networkClient.wakeup()
txnMarkerSendThread.awaitShutdown()

guozhangwang replied:
That's right. Ack.

junrao reviewed (commented):
@guozhangwang : Thanks for the updated patch. A few more comments.

hachikuji reviewed (approved):
Latest changes LGTM. Left one suggestion on the replica manager changes.

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K5231-read-write-lock

guozhangwang committed:
address Jun's comments round four

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala:
```diff
@@ -83,14 +93,24 @@ class TransactionMarkerRequestCompletionHandler(brokerId: Int,
-          case None =>
-            info(s"Transaction metadata for $transactionalId does not exist in the cache" +
-              s"any more; cancel sending transaction markers $txnMarker to the brokers")
+        val dummy = txnStateManager.getAndMaybeAddTransactionState(transactionalId)
```
It doesn't seem that we really need to define dummy?

guozhangwang committed:
fix unit tests

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -87,13 +85,16 @@ class TransactionStateManager(brokerId: Int,
   private val transactionTopicPartitionCount = getTransactionTopicPartitionCount
 
-  // this is best-effort expiration and hence not grabing the lock on metadata upon checking its state
+  // this is best-effort expiration and hence not grabbing the lock on the metadata object upon checking its state
```
Hmm, transactionsToExpire() should read the most up to date txnMetadata.txnStartTimestamp. So, we probably still need to synchronize on txnMetadata.

guozhangwang replied:
Ack. I will add this check after grabbing the lock inside `TxnCoordinator`.

junrao commented on core/src/main/scala/kafka/server/DelayedOperation.scala:
```diff
@@ -253,6 +253,9 @@ final class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: Stri
   def delayed: Int = timeoutTimer.size
 
+  /**
+    * Cancel watching on any delayed operations for the given key. Note the operation will not be completed
```
Since we are turning off reaper for DelayedTxnMarker, could we change tryCompleteElseWatch() to not add the operation to timeoutTimer if the reaper is disabled? Otherwise, we will be leaking memory.

guozhangwang replied:
Ack.

junrao reviewed (commented):
@guozhangwang : Thanks for the latest patch. LGTM. Just a few more minor comments. Once they are addressed, you can merge the patch.

guozhangwang commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -484,13 +468,45 @@ class TransactionStateManager(brokerId: Int,
-      isFromClient = false,
-      recordsPerPartition,
-      updateCacheCallback)
+    inReadLock(stateLock) {
```
cc @dguy

guozhangwang committed:
address Jun's comments round five

junrao commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -279,27 +262,31 @@ class TransactionCoordinator(brokerId: Int,
-    if (error != Errors.NONE)
-      responseCallback(error)
+                           responseCallback: EndTxnCallback,
+                           checkTxnTimeout: Boolean = false): Unit = {
```
This is ok, but it seems it's simpler to check the expiration time in one place in TransactionStateManager.transactionsToExpire().

guozhangwang replied:
Hmm, I thought it may be less efficient to traverse through the map and grab the lock on each object and check if it should be aborted, and hence I did this in two phases; plus by doing this we can also reuse the `handleEndTxn` function (note that we need to change its pending state while holding the lock), otherwise we need to do that in a separate function.

But let me know if you have a strong preference on this, then I can change it.

junrao replied:
My only concern is that handleEndTransaction() is now a bit more complicated to reason about. For example, if checkTxnTimeout is specified, should we validate that txnMarkerResult is always abort? 

A compromise is to make the timestamp in TransactionMetadata volatile. That way, the expiring method can get the up to date timestamp w/o a lock.

guozhangwang replied:
Alright.

junrao reviewed (commented):
@guozhangwang : LGTM. Just one minor comment.

guozhangwang committed:
fix unit tests

guozhangwang committed:
rebase from trunk

guozhangwang committed:
Jun's comment round six

guozhangwang committed:
minor fix

junrao commented:
@guozhangwang : Thanks for the latest patch. LGTM.

guozhangwang commented:
Merged to trunk and 0.10.0.

ijuma commented:
I think you mean 0.11.0.

guozhangwang commented:
> I think you mean 0.11.0.

Yup :)
